src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AwaitExpressionSpiller.cs

```diff
@@ -499,9 +506,17 @@ private BoundStatement UpdateStatement(BoundSpillSequenceBuilder builder, BoundS
                 }
             }
 
-            for (int i = lastSpill + 1; i < newList.Length; i++)
+            // the value of the last spill and everything that follows is not spilled
+            if (lastSpill < newList.Length)
```

Comment:
The use of `sideEffectsOnly` in this method seems inconsistent. Depending in the content of the `newList` when `sideEffectsOnly` is true we will return an empty list, a list which is part of `newList` or all of `newList`. That intuitively seems wrong. Am I missing something subtle here?

Reply:
`Spill` ensures that `replacement` is not sideeffecting (it is basically the "rehydrate" code). All sideeffects are moved into the spill builder. Therefore we can drop replacements when we know we only need sideeffects. - like if we are spilling a sequence. In fact some replacements are already `null` when coming from Spill - Spill knows when we do `sideeffectsOnly` and may just return `null` as the replacement as an easy-out when convenient. 

This is essentially an optimization, but it is an important one. 
If we keep unnecessary replacements, it will look like the values are spilled, survive across await and get restored - only to be unused later...

If we drop replacements right here, the spills become dead stores that do not need to be captured. Furthermore - IL Optimizer will kill dead stores and their locals, so only sideeffects will be emitted.

---
In reply to: [162102259](https://github.com/dotnet/roslyn/pull/24273#discussion_r162102259) [](ancestors = 162102259)

Reply:
At the same time though the method name is `VisitX` and it returns an array of varying length. Generally I would expect `VisitX` style methods to return an equivalent node. Maybe it's just the name throwing me off.

Reply:
I think the confusing part is that this is a bit of unusual rewrite. It is not purely node-to-node. We are also filling up a builder. Parts of the input may move into the builder so it may appear that we have lost something. 
Most of the VisitX methods in this rewriter are like that. 

---
In reply to: [162472242](https://github.com/dotnet/roslyn/pull/24273#discussion_r162472242) [](ancestors = 162472242)